File: src/main/java/no/fintlabs/cache/CompletedFullSyncConsumer.java
package no.fintlabs.cache;

import lombok.RequiredArgsConstructor;
import no.fintlabs.kafka.event.EventConsumerFactoryService;
import no.fintlabs.kafka.event.topic.EventTopicNameParameters;
import no.fintlabs.status.models.ResourceEvictionPayload;
import org.apache.kafka.clients.consumer.ConsumerRecord;
import org.springframework.context.annotation.Bean;
import org.springframework.kafka.listener.ConcurrentMessageListenerContainer;
import org.springframework.stereotype.Service;

@Service
@RequiredArgsConstructor
public class CompletedFullSyncConsumer {

    private final CacheEvictionService cacheEvictionService;
    private final EventTopicNameParameters eventTopicName = createEventTopicNameParameters();

    @Bean
    public ConcurrentMessageListenerContainer<String, ResourceEvictionPayload> completedFullSyncListener(EventConsumerFactoryService eventConsumerFactoryService) {
        return eventConsumerFactoryService.createFactory(ResourceEvictionPayload.class, this::consume)
                .createContainer(eventTopicName);
    }

    private void consume(ConsumerRecord<String, ResourceEvictionPayload> consumerRecord) {
        cacheEvictionService.triggerEviction(consumerRecord.value());
    }

    private EventTopicNameParameters createEventTopicNameParameters() {
        return EventTopicNameParameters.builder()
                .orgId("fintlabs-no")
                .domainContext("fint-core")
                .eventName("completed-full-sync")
                .build();
    }

}

File: src/main/java/no/fintlabs/cache/CompletedFullSyncConsumer.kt
package no.fintlabs.cache

import no.fintlabs.consumer.config.ConsumerConfiguration
import no.fintlabs.kafka.event.EventConsumerFactoryService
import no.fintlabs.kafka.event.topic.EventTopicNameParameters
import no.fintlabs.status.models.ResourceEvictionPayload
import org.apache.kafka.clients.consumer.ConsumerRecord
import org.springframework.context.annotation.Bean
import org.springframework.stereotype.Component

@Component
class CompletedFullSyncConsumer(
    private val consumerConfig: ConsumerConfiguration,
    private val cacheEvictionService: CacheEvictionService
) {

    @Bean
    fun completedFullSyncContainer(eventConsumerFactoryService: EventConsumerFactoryService) =
        eventConsumerFactoryService.createFactory(
            ResourceEvictionPayload::class.java,
            this::consume
        ).createContainer(createEventTopic())

    private fun consume(consumerRecord: ConsumerRecord<String?, ResourceEvictionPayload>) =
        consumerRecord.value().takeIf { consumerConfig.matchesConfiguration(it.domain, it.pkg, it.org) }
            ?.let { cacheEvictionService.triggerEviction(it) }

    private fun createEventTopic(): EventTopicNameParameters =
        EventTopicNameParameters.builder()
            .orgId("fintlabs-no")
            .domainContext("fint-core")
            .eventName("completed-full-sync")
            .build()
}